### DISCUSSION TITLE
Will Transmission 4.x be easily installable on Debian 12 and Raspberry Pi? [Might pre-compiled versions of 4.1.0-dev be made available daily—to encourage thoughtful testing?]

### DISCUSSION THREAD

**Discussion by holta** [transmission/transmission#5585] [Q&A] [answered]
### What is the issue?

Ubuntu 23.10 daily builds make it easy to install Transmission 4.0.2 which is great:

```
$ apt list transmission*
Listing... Done
transmission-cli/mantic 4.0.2-1ubuntu2 amd64
transmission-common/mantic 4.0.2-1ubuntu2 amd64
transmission-daemon/mantic 4.0.2-1ubuntu2 amd64
transmission-gtk/mantic 4.0.2-1ubuntu2 amd64
transmission-qt/mantic 4.0.2-1ubuntu2 amd64
transmission-remote-gtk/mantic 1.5.1-1 amd64
transmission/mantic 4.0.2-1ubuntu2 amd64
```

A huge thank you to whoever made the above happen!

Finally for the masses who are unable to compile — can similar install paths possibly be made available for Debian 12 and/or [Raspberry Pi OS](https://forums.raspberrypi.com/viewtopic.php?t=348757) ([staging](http://raspbian.raspberrypi.org/raspbian/dists/bookworm-staging/)) ?

Background: Debian 12 Bookworm is expected to be released on June 10th later this week.

### Which application of Transmission?

(All?)

### Which version of Transmission?

4.0.3+

### How this might help

If such install paths become easier, that would be incredibly useful for the [Internet-in-a-Box](https://internet-in-a-box.org) community here:

- iiab/iiab#1078
- iiab/iiab#3526

**Comment by killemov** [comment]
That is completely up to each distribution and I don't have any indication that these packages will become obsolete in the near future.

**Comment by holta** [comment]
Maybe I'm dreaming but let me ask:

What's the best place for pre-compiled versions of Transmission 4.x to be made broadly available — with succinct install explanation for less-technical users?

Reason: OS's like Debian 12 appear to be stuck with the 2020 version of Transmission 3.0 until 2025.  Not a wise decision (on their part!) but that is what it is — so given circumstances like this (waiting half a decade why exactly ;-) might there be some way to broadly & easily help make Transmission 4.x installs much more understandable + much more popular?

**Comment by killemov** [comment]
Ok. We disagree on the wisdom of distribution shot-callers. IF Debian did package the bleeding edge version of TM, you would have had to deal with a lot of bugs that seriously impact transmission functionality. So I expect Debian to incorporate proven software in 'testing' and then get hardened from that point on. So when TM gets to 4.1 or 4.0.99 :smiley: only then would I expect it in testing.

**Comment by holta** [comment]
Does Transmission 4.0.3 have a lot of rough edges?

I'm certainly not in a position to judge!

But it seems like compiled versions *could* greatly help ground-truthing <-> testing and/or helping everyone decide for themselves if 4.1 or 4.0.99 are perhaps more appropriate.  So it's just a suggestion towards trying to help the Transmission ecology / ecosystem.

(No matter what distributions decide, typically years down the road!)

**Comment by killemov** [comment]
I get it but specifically Debian is about tried and true software. And they have made mistakes in the past by letting stuff through and then scrambling to get it fixed under big loads of bug reports. There are many distros, OpenSUSE Tumbleweed for example, that accept the issues with running newer software because of a different user base who accept this as well.

**Comment by ckerr** [comment]
Converting to a discussion since this isn't an issue that the Transmission project can resolve, but for sure it's worth a discussion here where other people can find it.

> Does Transmission 4.0.3 have a lot of rough edges?

The biggest issue that I've heard people complain about is that it uses a newer version of GTK and themes haven't quite caught up, so people using it with some GTK themes have noticed that it doesn't pick up their system's theme.

**Comment by ile6695** [comment]
Debian does already have 4.x inside their [GitLab repository](https://salsa.debian.org/debian/transmission), which means it will likely appear in their unstable branch after Debian 12 has been released. Hopefully backporting from there will  be easy.

The [docs](https://github.com/transmission/transmission/tree/main/docs) already have instructions for building transmission [for Debian 11](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#debian-11--bullseye) and there will later be similar instructions for 12.

**Comment by holta** [comment]
To encourage regular and thoughtful testing, another option might be for GitHub Actions to auto-compile (timestamped) test builds after every commit landing on Transmission's [main branch](https://github.com/transmission/transmission/tree/main) — as [yt-dlp](https://github.com/yt-dlp/yt-dlp) does for their master branch here, in a secondary repo they've nicknamed "nightly": (probably not the best name!)

https://github.com/yt-dlp/yt-dlp-nightly-builds/releases

If so a couple common compile targets for Linux (e.g. x86_64 and aarch64) might make sense?  To help nurture/broaden responsive functional testing on many/most Linux distributions.

**Comment by GaryElshaw** [comment]
I'm not sure if this is useful, but you can try downloading artefacts from commits. For example, from the latest commit, if you scroll down, a debian package is there: https://github.com/transmission/transmission/actions/runs/5180924436

**Comment by holta** [comment]
Many thanks @GaryElshaw for pointing that out!

Those auto-generated zipfiles [binaries-debian-11-from-tarball](https://github.com/transmission/transmission/suites/13386844392/artifacts/732826748) and [binaries-ubuntu-20-04-from-tarball](https://github.com/transmission/transmission/suites/13386844392/artifacts/732826754) each contain a single ~3 MB binary (transmission-qt) — without tools like `transmission-daemon` and `transmission-remote` etc regrettably, i.e. bare bones as you suggest — _but at least a start!_

PS for now build targets include EOL'd Fedora 36 and an older version of Ubuntu LTS (20.04 instead of 22.04).  Would be really great to update such targets in 2023.  If the stars align, ideally also using https://github.com/actions/upload-artifact to upload a couple more Transmission artifacts for broader community testing of 4.1.0-dev 🌠

**Reply by GaryElshaw** [reply]
Excellent suggestion here @mikedld and @ckerr.

**Comment by neomatrixcode** [comment]
Good day, I understand that the github action "actions.yml" creates a compilation every time a change is made to any of the parts of the library (make-cli, make-daemon, make-qt, etc), but it will be possible to include permanently to make-cli and make-daemon on each of those builds so I can download and test these nightly builds?

```
jobs:
  what-to-make:
    runs-on: ubuntu-22.04
    outputs:
      make-cli: ${{ steps.check-main-push.outputs.is-main-push == '1' || steps.check-diffs.outputs.cli-changed == '1' }}
      make-daemon: ${{ steps.check-main-push.outputs.is-main-push == '1' || steps.check-diffs.outputs.daemon-changed == '1' }}
```

```
jobs:
  what-to-make:
    runs-on: ubuntu-22.04
    outputs:
      make-cli: True
      make-daemon: True
```

**Reply by holta** [reply]
Interesting idea!💡

Current code's 10+ make targets:

https://github.com/transmission/transmission/blob/0ef58c2a20019096aed0ab639f2207b8e1c1b248/.github/workflows/actions.yml#L15-L26

Based on some clever `git diff` tests just below:

https://github.com/transmission/transmission/blob/0ef58c2a20019096aed0ab639f2207b8e1c1b248/.github/workflows/actions.yml#L56-L67

Can someone suggest a way forward — towards [auto-uploading a few more artifacts](https://github.com/actions/upload-artifact) for thoughtful and timely CI/CD-driven community testing?

**Reply by holta** [reply]
Just FYI GitHub [removes uploaded artifacts](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#artifact-and-log-retention-policy
) after 90 days:

![image](https://github.com/transmission/transmission/assets/2458907/4bf6a914-f38a-4754-89d9-4c6f6fc172c0)

90 days appears to be plenty for community testing?

(Though some people will of course be caught by surprise not realizing this!)

Possibly this is the reason yt-dlp chose to post "permanent" nightly-if-not-hourly builds to a separate repo at: https://github.com/yt-dlp/yt-dlp-nightly-builds/releases

**Comment by evrial** [comment]
So whats the status, nobody cares? Most actual raspberry pi OS (Debian 12) has 3 year old arm64 binaries.
https://packages.debian.org/ru/bookworm/transmission-daemon

**Reply by nevack** [reply]
Who exactly do you want to care about this?

**Reply by evrial** [reply]
Ok whatever, installed from Debian sid
```
transmission-daemon -V
transmission-daemon 4.0.2 (2a57b17031)
```

**Reply by nevack** [reply]
If you want to see this version back-ported to `bookworm` or any other release - just ask Debian maintainers. Contact information is listed on every package's page.

**Reply by evrial** [reply]
https://qa.debian.org/excuses.php?package=transmission
https://bugs.debian.org/cgi-bin/pkgreport.cgi?src=transmission
```
Outstanding bugs -- Serious (policy violations or makes package unfit for release); Unclassified (2 bugs)
[#1041519](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1041519) [S|  |  ] [[src:transmission](https://bugs.debian.org/cgi-bin/pkgreport.cgi?src=transmission)] [transmission: contains prebuilt javascript without source](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1041519)
[#1042027](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1042027) [S|⛺|  ] [[src:transmission](https://bugs.debian.org/cgi-bin/pkgreport.cgi?src=transmission)] [transmission: FTBFS: Utils.h:125:6: error: variable or field ‘gtr_priority_combo_init’ declared void](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1042027)
```
Not sure if fixed in 4.0.4 or not

**Reply by evrial** [reply]
I guess minified JS in repository is no-go for Debian policy

**Reply by evrial** [reply]
My life is too short for this garbage

**Reply by tearfur** [reply]
@ckerr @mikedld It looks like Debian [#1041519](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1041519) means that the way we build the minified JS using the CI is violating Debian policies. I think this is important and it is best addressed by the lead developers.

**Reply by ckerr** [reply]
@tearfur not sure what we could do here upstream? It's not as if we can bundle node + npm + the 143M in `node_modules` in our tarballs.

And even if we _did_ encourage users to build from source, that would limit which platforms Transmission could be built on, because older distros don't have easy access to modern versions of npm.

I totally understand why Debian would prefer to build from source instead of trusting that the generated files are safe. This is the same reason that we have bots that update transmission-web.js automatically and which refuse PRs that change that file. If there's something we can do to make life easier for the Debian packagers, I'm happy to listen & happy to help. The current setup was chosen because it was the only option that I could find that let people build transmission from tarball on all platforms, not because I want to make anyone's life difficult.

**Reply by evrial** [reply]
You could use modern esbuild which has no extra dependencies and 100x faster than babel and webpack. Or you could simply not use asset minification exactly as you did in 3 version

```
apt install esbuild
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Recommended packages:
  nodejs
The following NEW packages will be installed:
  esbuild
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 4947 kB of archives.
After this operation, 20.3 MB of additional disk space will be used.
Get:1 https://deb.debian.org/debian bookworm/main arm64 esbuild arm64 0.17.0-1+b2 [4947 kB]
Fetched 4947 kB in 1s (6159 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package esbuild.
(Reading database ... 35545 files and directories currently installed.)
Preparing to unpack .../esbuild_0.17.0-1+b2_arm64.deb ...
Unpacking esbuild (0.17.0-1+b2) ...
Setting up esbuild (0.17.0-1+b2) ...
```

**Reply by niol** [reply]
I'm looking into getting transmission 4.0 webapp into Debian.
Debian policy compliance makes building from source mandatory, with no network connection.

Bundling the node_modules directory would not help because all dependencies again need to be built from source and some are already packaged in Debian, and Debian tries to avoid duplication.

What the transmissionbt developpers can do to make the Debian packagers life easier is lower the number of webapp build dependencies, maybe in a specific build target. For instance of what needs to be adressed on the Debian side:

- ```css-minimizer-webpack-plugin``` is not available in Debian. Maybe it is acceptable for CSS to not be minimized rather than spending time packaging this.
- ```sass-loader``` is not available in Debian. I guess this cannot be worked around.

**Reply by evrial** [reply]
Again, we had zero issues in 3 version until transmission author decided to use source code obfuscation. Think about that.
Saving some kilobytes at cost of security issues. If you ask me, this web mess should be scrapped and /web/ restored to prior v3.

**Reply by niol** [reply]
I completely understand what makes language package managers (`npm` but also `pip`, `cargo`, `composer`, `gem` and many more) appealing for development environments. Unfortunately, that makes building those apps less resilient to service outage, service change, and more difficult to manage regarding the external package repositories moving target, which is not acceptable for some distributions.

Regarding that specific issue with the transmission webapp, I managed to have it build without `npm` in a Debian environment. [My pull request regarding building the webapp in the Debian package](https://salsa.debian.org/debian/transmission/-/merge_requests/16) uses dependencies available in Debian, builds using only `cd web; webpack --config webpack.config.js` and makes [the following changes to the Transmission source](https://salsa.debian.org/debian/transmission/-/merge_requests/16/diffs#73aaf45f6b6a2be5b95b67c8ef2912c0f831edee) to ease the work on the Debian side:
- do not minimize CSS
- do not use sass `math.*` functions that require `dart-sass`

This will hopefully make its way into the Debian repositories when the maintainer will have time to have a look. I also have builds.

**Reply by evrial** [reply]
How do you get webpack on the system? You need npm. Honestly I believe @ckerr should stop and rethink what is transmission-daemon is about before we suddenly got 3GB of electron freedom and democracy shipped with it.

**Reply by niol** [reply]
`webpack` is available in Debian, so `sudo apt install webpack`. Again, the build I have proposed is self contained, does not use `npm` and builds in a clean chroot.

**Reply by evrial** [reply]
```
 apt install webpack
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  eslint handlebars libc-ares2 libjs-async libjs-events libjs-inherits libjs-is-typedarray libjs-prettify
  libjs-regenerate libjs-source-map libjs-sprintf-js libjs-typedarray-to-buffer libjs-util libnode108 node-abbrev
  node-acorn node-ajv node-ajv-keywords node-ampproject-remapping node-ansi-escapes node-ansi-regex node-ansi-styles
  node-anymatch node-archy node-argparse node-arrify node-assert node-async node-async-each
  node-babel-helper-define-polyfill-provider node-babel-plugin-add-module-exports node-babel-plugin-lodash
  node-babel-plugin-polyfill-corejs2 node-babel-plugin-polyfill-corejs3 node-babel-plugin-polyfill-regenerator
  node-babel7 node-babel7-runtime node-balanced-match node-binary-extensions node-brace-expansion node-braces
  node-browserslist node-busboy node-camelcase node-caniuse-lite node-chalk node-chokidar node-chrome-trace-event
  node-ci-info node-cjs-module-lexer node-cliui node-clone node-clone-deep node-color-convert node-color-name
  node-commander node-commondir node-concat-stream node-convert-source-map node-core-js node-core-js-compat
  node-core-js-pure node-core-util-is node-debbundle-es-to-primitive node-debug node-decamelize node-deep-equal
  node-deep-is node-defaults node-define-properties node-defined node-del node-doctrine node-electron-to-chromium
  node-enhanced-resolve node-errno node-error-ex node-es-abstract node-es-module-lexer node-es6-error
  node-escape-string-regexp node-escodegen node-eslint-scope node-eslint-utils node-eslint-visitor-keys node-espree
  node-esprima node-esquery node-esrecurse node-estraverse node-esutils node-events node-fast-deep-equal
  node-fast-levenshtein node-file-entry-cache node-fill-range node-find-cache-dir node-find-up node-flat-cache
  node-flatted node-for-in node-for-own node-foreground-child node-fs-readdir-recursive node-fs.realpath
  node-function-bind node-functional-red-black-tree node-get-caller-file node-glob node-glob-parent node-globals
  node-globby node-graceful-fs node-has-flag node-hosted-git-info node-ignore node-imurmurhash node-indent-string
  node-inflight node-inherits node-interpret node-is-arrayish node-is-binary-path node-is-buffer node-is-extendable
  node-is-extglob node-is-glob node-is-number node-is-path-cwd node-is-path-inside node-is-plain-obj
  node-is-plain-object node-is-stream node-is-typedarray node-is-windows node-isarray node-isexe node-isobject
  node-istanbul node-jest-debbundle node-jest-worker node-js-tokens node-js-yaml node-jsesc
  node-json-parse-better-errors node-json-schema node-json-schema-traverse node-json-stable-stringify node-json5
  node-jsonify node-kind-of node-levn node-loader-runner node-locate-path node-lodash node-lodash-packages
  node-lru-cache node-make-dir node-memfs node-memory-fs node-merge-stream node-micromatch node-mime node-mime-types
  node-minimatch node-minimist node-mkdirp node-ms node-n3 node-neo-async node-nopt node-normalize-package-data
  node-normalize-path node-npm-run-path node-object-assign node-object-inspect node-once node-optimist node-optionator
  node-p-limit node-p-locate node-p-map node-parse-json node-path-dirname node-path-exists node-path-is-absolute
  node-path-is-inside node-path-type node-picocolors node-pify node-pkg-dir node-postcss node-prelude-ls
  node-process-nextick-args node-progress node-prr node-punycode node-quick-lru node-randombytes node-read-pkg
  node-readable-stream node-readdirp node-rechoir node-regenerate node-regenerate-unicode-properties
  node-regenerator-runtime node-regenerator-transform node-regexpp node-regexpu-core node-regjsgen node-regjsparser
  node-repeat-string node-require-directory node-resolve node-resolve-cwd node-resolve-from node-resumer node-rimraf
  node-safe-buffer node-schema-utils node-semver node-serialize-javascript node-set-immediate-shim
  node-shebang-command node-shebang-regex node-signal-exit node-slash node-slice-ansi node-source-list-map
  node-source-map node-source-map-support node-spdx-correct node-spdx-exceptions node-spdx-expression-parse
  node-spdx-license-ids node-sprintf-js node-string-decoder node-string-width node-strip-ansi node-strip-bom
  node-strip-json-comments node-supports-color node-tapable node-tape node-terser node-through node-to-fast-properties
  node-to-regex-range node-tslib node-type-check node-typedarray node-typedarray-to-buffer node-undici
  node-unicode-canonical-property-names-ecmascript node-unicode-match-property-ecmascript
  node-unicode-match-property-value-ecmascript node-unicode-property-aliases-ecmascript node-uri-js node-util
  node-util-deprecate node-uuid node-v8-compile-cache node-v8flags node-validate-npm-package-license node-watchpack
  node-wcwidth.js node-webassemblyjs node-webpack-sources node-which node-wordwrap node-wrap-ansi node-wrappy
  node-write node-write-file-atomic node-xtend node-y18n node-yallist node-yargs node-yargs-parser nodejs terser
Suggested packages:
  node-babel-eslint node-esprima-fb node-inquirer libjs-angularjs node-babel-plugin-polyfill-es-shims
  node-babel7-debug javascript-common livescript chai node-jest-diff npm
Recommended packages:
  node-text-table javascript-common nodejs-doc
^C
```

**Reply by tearfur** [reply]
These are a lot, but aren't they all just build dependencies? What's the big deal then?

**Reply by ckerr** [reply]
- I really like @evrial's suggestion of migrating from webpack to esbuild and have put up a PR at https://github.com/transmission/transmission/pull/6280 to do this.

- There doesn't seem to be any good way to have [`optionalDevDependencies` in npm](https://github.com/npm/npm/issues/3870) which would be handy here to separate the build dependencies (e.g. esbuild/webpack) from the development dependencies (e.g. eslint, prettier). So #6280 also adds a `package.json.buildonly` file that packagers can overwrite `package.json` with to have _only_ the dependencies needed for building. 

- 6280 addresses some of @niol's [comments](https://github.com/transmission/transmission/discussions/5585#discussioncomment-7564933) about what's blocking on the Debian side by removing those dependencies. There is still followup action needed for `esbuild-sass-plugin` which is tracked at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1051470, but at least this _reduces_ the number of pain points. 

- @niol thank you for your work in the Debian PR! PTAL at #6280, I'm happy to make any changes that can simplify this further for Debian packagers.

- @evrial I appreciate your comments here, especially the esbuild suggestion, but please keep discussion on a productive level. Comments like "until transmission author decided to use source code obfuscation" (implying that I'm trying to hide something in the bundles) or "3GB of electron freedom and democracy" (I am an Electron maintainer) are not productive and are not appreciated.

**Comment by Nemric** [comment]
Perhaps should you try this : https://github.com/transmission/transmission/discussions/5710
I don't make any arm build but you can make your own based on my Containerfile ?

**Comment by holta** [comment]
Debian 12 (documentation) progress evolving here:

- PR #5866

**Comment by ckerr** [comment]
f7f6ff23e219ad9840a7274d685953e72662bc03 backports the `README.md` instructions to `4.0.x`, so it will be in the upcoming 4.0.5 release.

tldr this includes simple instructions on how to build Transmission on Debian systems using only Debian packages (esbuild, perl, rsass) instead of `Node.js` so that no `npm install` step is necessary.

**Comment by ckerr** [comment, marked as answer]
4.0.5 has been released @ https://github.com/transmission/transmission/releases/tag/4.0.5 and its web client is buildable without Node.js and using only apt-installed Debian packages as described in the `web/README.md`.

So let's hope this makes life easier for the Debian packagers

**Reply by ckerr** [reply]
CC @niol ^

**Reply by niol** [reply]
I'm in touch with the Debian maintainer and things will hopefully move forward over the week-end. There aren't any technical blockers that I can identify. I have a Debian policy compliant build working.

**Reply by zerlas** [reply]
will it be available anytime soon on bookworm ?

**Reply by niol** [reply]
I have had no answer from the maintainers so this will take more time, unfortunately.

**Reply by TheFireCookie** [reply]
Any update on this topic @niol  ? :)

**Reply by niol** [reply]
transmission is in good shape in Debian. Regarding the bookworm backport, I'm searching for a sponsor to make it available but there are no technical showstoppers (I use a self-built backport myself with no issue).

**Reply by reinisb** [reply]
It is November 2024 and I still get transmission v3 on the latest Raspberry Pi OS version. It's a shame there are no instructions how to build transmission v4.

```
pi@rpi5:~ $ apt list transmission*
Listing... Done
transmission-cli/stable,now 3.00-2.1+deb12u1 arm64 [installed,automatic]
transmission-cli/stable 3.00-2.1+deb12u1 armhf
transmission-common/stable,stable,now 3.00-2.1+deb12u1 all [installed,automatic]
transmission-daemon/stable,now 3.00-2.1+deb12u1 arm64 [installed]
transmission-daemon/stable 3.00-2.1+deb12u1 armhf
transmission-gtk/stable 3.00-2.1+deb12u1 arm64
transmission-gtk/stable 3.00-2.1+deb12u1 armhf
transmission-qt/stable 3.00-2.1+deb12u1 arm64
transmission-qt/stable 3.00-2.1+deb12u1 armhf
transmission-remote-gtk/stable 1.5.1-1 arm64
transmission-remote-gtk/stable 1.5.1-1 armhf
transmission/stable,stable 3.00-2.1+deb12u1 all
```

**Reply by reinisb** [reply]
Are [these](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#debian-12--bookworm) the correct instructions?

**Reply by ile6695** [reply]
@reinisb having v3 in there is expected, because those distros have their own update schedules. Have you tried searching for build instructions (in this very page) or are you saying there are none in hopes that someone else will do that work for you?

**Reply by ile6695** [reply]
> Are [these](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#debian-12--bookworm) the correct instructions?

Those are indeed instructions that will most likely work for raspberry.

**Comment by zerlas** [comment]
Oh, that's unfortunate... I sended a message to maintainers too, maybe it will be published soon. Many peoples are waiting for this package on stablesLe 29 mars 2024 22:10, niol ***@***.***> a écrit :
I have had no answer from the maintainers so this will take more time, unfortunately.
